src/intent_parser/lookup_intent_parser.rs

```diff
@@ -0,0 +1,824 @@
+        let entities = self.get_all_entities(input)?;
+        let (_, formatted_input) =
+            replace_entities(input, entities.clone(), get_entity_placeholder);
+        let cleaned_input = self.preprocess_text(input);
+        let cleaned_formatted_input = self.preprocess_text(&*formatted_input);
+        let key = self.generate_key(&cleaned_formatted_input);
+        let val = if let Some(v) = self.map.get(&key) {
+            Some(v)
+        } else {
+            self.map.get(&cleaned_input.to_lowercase())
+        };
+
+        Ok(self.parse_map_output(input, val, entities, intents_whitelist))
+    }
+
+    fn get_intents(&self, input: &str) -> Result<Vec<IntentClassifierResult>> {
+        let mut intents = vec![];
+        let res = self.parse(input, None)?;
+        let names = if let Some(ref name) = res.intent.intent_name {
+            intents.push(res.intent.clone());
+            self.intents_names
+                .iter()
+                .filter(|x| *x != name)
+                .cloned()
+                .collect::<Vec<_>>()
+        } else {
+            self.intents_names.clone()
+        };
+        for name in names {
+            intents.push(IntentClassifierResult {
+                intent_name: Some(name),
+                probability: 0.0,
+            })
+        }
+        intents.push(IntentClassifierResult {
+            intent_name: None,
+            probability: 0.0,
+        });
+
+        Ok(intents)
+    }
+
+    fn get_slots(&self, input: &str, intent: &str) -> Result<Vec<InternalSlot>> {
+        if !self.intents_names.contains(&intent.to_string()) {
+            return Err(SnipsNluError::UnknownIntent(intent.to_string()).into());
+        }
+        let filter = vec![intent];
+        self.parse(input, Some(&filter)).map(|result| result.slots)
+    }
+}
+
+impl LookupIntentParser {
+    fn preprocess_text(&self, string: &str) -> String {
+        let tokens = tokenize(string, NluUtilsLanguage::from_language(self.language));
+        let mut current_idx = 0;
+        let mut cleaned_string = "".to_string();
+        for mut token in tokens {
+            if self.ignore_stop_words
+                && self
+                    .shared_resources
+                    .stop_words
+                    .contains(&token.normalized_value())
+            {
+                token.value = (0..token.value.chars().count()).map(|_| " ").collect();
+            }
+            let prefix_length = token.char_range.start - current_idx;
+            let prefix: String = (0..prefix_length).map(|_| " ").collect();
+            cleaned_string = format!("{}{}{}", cleaned_string, prefix, token.value);
+            current_idx = token.char_range.end;
+        }
+        let suffix_length = string.chars().count() - current_idx;
+        let suffix: String = (0..suffix_length).map(|_| " ").collect();
+        cleaned_string = format!("{}{}", cleaned_string, suffix);
+        cleaned_string
+    }
```

Comment:
This method can be simplified as we don't need to ensure that the processed string have the same length as the original one.